### PR TITLE
WIP: Add JSON validator

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | yes
| Fixed issues  | #15524

### CODE REVIEW

File: framework/helpers/BaseJson.php [hunk 1]
```diff
@@ -109,6 +109,31 @@ public static function decode($json, $asArray = true)
...
+     * @return boolean `true` if given data is a valid JSON string, `false` otherwise.
+     *
+     * @since 2.0.14
+     */
+    public static function validate($data)
+    {
+        $isValid = false;
+
+        if (is_string($data) === true) {
+            json_decode($data);
```
Comment (SilverFire) [inline framework/helpers/BaseJson.php]: I'd add the STFU operator just for the case.

**Review by SilverFire — APPROVED** [review]

File: framework/helpers/BaseJson.php (hunk 1 above)
Comment (samdark) [reply framework/helpers/BaseJson.php]: Nope. It doesn't raise any errors.

File: framework/helpers/BaseJson.php [hunk 2]
```diff
@@ -109,6 +109,31 @@ public static function decode($json, $asArray = true)
...
+
+            $lastError = json_last_error();
+            if ($lastError === JSON_ERROR_NONE) {
```
Comment (samdark) [inline framework/helpers/BaseJson.php]: This info is valuable so I'd think twice before returning just true/false and losing it.

Comment (samdark) [reply framework/helpers/BaseJson.php]: http://www.beliefmedia.com/code/php-snippets/check-valid-json

**Comment by klimov-paul** [comment]
I am afraid that simple syntax check is not enough for JSON validation.
It should support matching particular internal structure as well. For example: there should be a way to determine whether some JSON can suite as a configuration for Composer.

**Comment by samdark** [comment]
Yep. That as well: https://github.com/Seldaek/jsonlint

**Comment by samdark** [comment]
And https://github.com/thephpleague/json-guard

**Comment by samdark** [comment]
Composer itself uses https://github.com/Seldaek/json-schema

File: framework/helpers/BaseJson.php (hunk 2 above)
Comment (Kolyunya) [reply framework/helpers/BaseJson.php]: @samdark I refactored the method so that now it returns actual validation errors.

**Comment by Kolyunya** [comment]
@klimov-paul @alekson described a use-case when he only needs to know if provided data is a valid JSON or not. We may add schema validation later on. These are two independent features.

File: framework/helpers/BaseJson.php (hunk 2 above)
Comment (Kolyunya) [reply framework/helpers/BaseJson.php]: @samdark another option is to return the error via the output parameter. In fact I like this option more.

**Comment by rob006** [comment]
Why did you add this to helper instead of create separate validator class?

**Comment by Kolyunya** [comment]
I've added this method to a helper for a convenience of use. In some scenarios you want to validate a value w/o instantiating a validator class. I think this method doesn't lower the helper cohesion.

I've also added a separate validator which makes use of a new helper method. But it's more like a proof of concept since I feel like the whole PR is going to be rejected.

**Comment by samdark** [comment]
I'm leaning towards rejecting it... While it may be useful, validating just JSON syntax doesn't sounds like a common requirement while validating schema doesn't sound like we should implement it (there are many good libraries available).

**Comment by Kolyunya** [comment]
@samdark what do you think about implementing a JSON schema validator based on existing solutions? This way we will provide a convenient tool out of box which works together with Yii forms w/o reimplementing complex logic.

**Comment by samdark** [comment]
What kind of Yii bindings do you propose? Validator?

**Comment by Kolyunya** [comment]
@samdark yes, I meant a validator.

**Comment by samdark** [comment]
I don't think we want to add another dependency to the core. btw., it's already available as extension: https://github.com/dstotijn/yii2-json-schema-validator


@Kolyunya thanks for contribution but it won't be accepted. Of course, that's me speaking. If other core team members disagree, we'll reopen it.

**Comment by alekson** [comment]
Жаль, хотелось простой валидатор на уровне numeric, string и т.д. - json, но который проверит что строка без ошибок конвертится в json.  Актуально для небольшого конфига, где не нужна сложная json schema.

Сейчас пользуюсь - https://github.com/romi45/yii2-json-validator, для  такой простой задачи не оч хотелось тянуть зависимость.

**Pull request closed without merging**